# Работа с SQL

Ваша компания купила крупный сервис для чтения книг по подписке. Необходимо проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Задания" data-toc-modified-id="Задания-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Задания</a></span></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Описание данных</a></span></li><li><span><a href="#Импорт-библиотек.-Подключение-к-БД.-Изучение-данных" data-toc-modified-id="Импорт-библиотек.-Подключение-к-БД.-Изучение-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Импорт библиотек. Подключение к БД. Изучение данных</a></span></li><li><span><a href="#Задание-№1" data-toc-modified-id="Задание-№1-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Задание №1</a></span></li><li><span><a href="#Задание-№2" data-toc-modified-id="Задание-№2-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Задание №2</a></span></li><li><span><a href="#Задание-№3" data-toc-modified-id="Задание-№3-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Задание №3</a></span></li><li><span><a href="#Задание-№4" data-toc-modified-id="Задание-№4-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Задание №4</a></span></li><li><span><a href="#Задание-№5" data-toc-modified-id="Задание-№5-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Задание №5</a></span></li></ul></div>

## Задания

1.	Посчитайте, сколько книг вышло после 1 января 2000 года;
2.	Для каждой книги посчитайте количество обзоров и среднюю оценку;
3.	Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
4.	Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
5.	Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

## Описание данных

**Таблица books**

Содержит данные о книгах:
-	book_id — идентификатор книги;
-	author_id — идентификатор автора;
-	title — название книги;
-	num_pages — количество страниц;
-	publication_date — дата публикации книги;
-	publisher_id — идентификатор издателя.

**Таблица authors**

Содержит данные об авторах:
-	author_id — идентификатор автора;
-	author — имя автора.

**Таблица publishers**

Содержит данные об издательствах:
-	publisher_id — идентификатор издательства;
-	publisher — название издательства;

**Таблица ratings**

Содержит данные о пользовательских оценках книг:
-	rating_id — идентификатор оценки;
-	book_id — идентификатор книги;
-	username — имя пользователя, оставившего оценку;
-	rating — оценка книги.

**Таблица reviews**

Содержит данные о пользовательских обзорах на книги:
-	review_id — идентификатор обзора;
-	book_id — идентификатор книги;
-	username — имя пользователя, написавшего обзор;
-	text — текст обзора.

<div class="alert alert-block alert-success">
    

__Комментарий от ревьюера №1 (OK)__ 

Хорошее начало проекта с заголовком, описанием и интерактивным меню.
</div>

## Импорт библиотек. Подключение к БД. Изучение данных 

In [ ]:
import pandas as pd
from sqlalchemy import create_engine
import datetime as dt

In [ ]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', 
 'pwd': 'Sdf4$2;d-d30pp', 
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 


**Выведем все таблицы**

In [ ]:
for table in ['books', 'authors', 'publishers', 'ratings', 'reviews']:
    query = ''' SELECT * FROM {}; 
    '''.format(table) 

    request = pd.io.sql.read_sql(query, con = engine) 

    print(f'Таблица - "{table}"')
    display(request.head())
    print(request.info())
    print('*********************************')

Таблица - "books"


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
None
*********************************
Таблица - "authors"


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
None
*********************************
Таблица - "publishers"


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
None
*********************************
Таблица - "ratings"


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
None
*********************************
Таблица - "reviews"


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
None
*********************************


**Данные в порядке, пропусков нет**

Запишем каждую таблицу в отдельные датафреймы, чтобы проверять наши запросы

In [ ]:
query_books = ''' SELECT * FROM books; ''' 
books = pd.io.sql.read_sql(query_books, con = engine)

query_reviews = ''' SELECT * FROM reviews; ''' 
reviews = pd.io.sql.read_sql(query_reviews, con = engine)

query_ratings = ''' SELECT * FROM ratings; ''' 
ratings = pd.io.sql.read_sql(query_ratings, con = engine)

query_publishers = ''' SELECT * FROM publishers; ''' 
publishers = pd.io.sql.read_sql(query_publishers, con = engine)

query_authors = ''' SELECT * FROM authors; ''' 
authors = pd.io.sql.read_sql(query_authors, con = engine)


## Задание №1 

**Посчитайте, сколько книг вышло после 1 января 2000 года**

In [ ]:
query_1 = ''' 
SELECT COUNT(*) 
FROM books
WHERE publication_date > '2000-1-1'; 
''' 

request_1 = pd.io.sql.read_sql(query_1, con = engine)


print(f'После 1 января 2000 года вышло {request_1.iloc[0,0]} книг')

После 1 января 2000 года вышло 819 книг


Проверка средствами Python

In [ ]:
books['publication_date'] = books['publication_date'].astype('datetime64')
books[books['publication_date']>'2000-01-01']['book_id'].count()

819

**Вывод: После 1 января 2000 года вышло 819 книг**

[*Назад к содержанию*](#Содержание)

## Задание №2 

**Для каждой книги посчитайте количество обзоров и среднюю оценку**

In [ ]:
query_2 = ''' 

SELECT 
      books.book_id,    
      books.title,
      avg_r.avg_rating,
      count_rev.reviews
FROM books
LEFT JOIN (SELECT book_id,
                  AVG(rating) avg_rating
           FROM ratings
           GROUP BY book_id
                ) AS avg_r ON avg_r.book_id = books.book_id
LEFT JOIN (SELECT book_id,
                  COUNT(review_id) AS reviews
                FROM reviews
                GROUP BY book_id
                ) AS count_rev ON count_rev.book_id = books.book_id
                
; ''' 

request_2 = pd.io.sql.read_sql(query_2, con = engine)
request_2

book_id                                              title  avg_rating  \
0        652          The Body in the Library (Miss Marple  #3)    4.500000   
1        273                                          Galápagos    4.500000   
2         51                           A Tree Grows in Brooklyn    4.250000   
3        951  Undaunted Courage: The Pioneering First Missio...    4.000000   
4        839                                        The Prophet    4.285714   
..       ...                                                ...         ...   
995       64                                Alice in Wonderland    4.230769   
996       55          A Woman of Substance (Emma Harte Saga #1)    5.000000   
997      148                                          Christine    3.428571   
998      790  The Magicians' Guild (Black Magician Trilogy  #1)    3.500000   
999      828                           The Plot Against America    3.000000   

     reviews  
0        2.0  
1        2.0  
2        5.0  
3        2.0  
4        4.0  
..       ...  
995      4.0  
996      2.0  
997      3.0  
998      2.0  
999      2.0  

[1000 rows x 4 columns]

[*Назад к содержанию*](#Содержание)

## Задание №3 

**Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры**

In [ ]:
query_3 = ''' 
SELECT
    pub.publisher_id,
    pub.publisher,
    pub.count_books
FROM
    (SELECT 
           books.publisher_id AS publisher_id,
           publishers.publisher AS publisher,
           COUNT(books.book_id) AS count_books
    FROM books 
    LEFT JOIN publishers ON books.publisher_id=publishers.publisher_id
    WHERE num_pages > 50
    GROUP BY books.publisher_id, publishers.publisher
    ORDER BY count_books DESC) AS pub
LIMIT 1
; ''' 

request_3 = pd.io.sql.read_sql(query_3, con = engine)
request_3

publisher_id      publisher  count_books
0           212  Penguin Books           42

Проверка средствами Python

In [ ]:
books.query('num_pages > 50')[['publisher_id',
                             'num_pages', 
                             'book_id']].groupby('publisher_id')['book_id'].count().reset_index().sort_values(by='book_id',ascending=False)


publisher_id  book_id
207           212       42
302           309       31
111           116       25
212           217       24
31             33       19
..            ...      ...
83             86        1
199           204        1
200           205        1
82             85        1
333           340        1

[334 rows x 2 columns]

In [ ]:
publishers.query('publisher_id == 212')

publisher_id      publisher
211           212  Penguin Books

**Вывод: Издательство, которое выпустило наибольшее число книг, толще 50 страниц - это "Penguin Books" выпустили 42 книги**

[*Назад к содержанию*](#Содержание)

## Задание №4 

**Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками**

In [ ]:
query_4 = ''' 
SELECT 
    a_avg.author,
    AVG(a_avg.avg_rating) AS avg_avg_rating
FROM
    (SELECT 
           authors.author AS author,
           AVG(ratings.rating) AS avg_rating
     FROM books
     LEFT JOIN authors ON books.author_id = authors.author_id
     LEFT JOIN ratings ON books.book_id = ratings.book_id
     GROUP BY books.book_id, authors.author_id
     HAVING COUNT(ratings.rating_id) >= 50) AS a_avg 
GROUP BY a_avg.author
ORDER BY avg_avg_rating DESC
LIMIT 1
    
; ''' 

request_4 = pd.io.sql.read_sql(query_4, con = engine)
request_4


author  avg_avg_rating
0  J.K. Rowling/Mary GrandPré        4.283844

**Вывод: Самую высокую оценку получает - J.K. Rowling/Mary GrandPré - 4.283844. Все любят Гарри Поттера** 

[*Назад к содержанию*](#Содержание)

## Задание №5 

**Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок**

In [ ]:
query_5 = ''' 
SELECT
      AVG(rev.count_reviews) AS avg_review
FROM
    (SELECT
           username,
           COUNT(review_id) AS count_reviews
    FROM reviews
    WHERE username IN (SELECT username
                       FROM ratings
                       GROUP BY username
                       HAVING COUNT(rating_id) > 50)
    GROUP BY  username) AS rev
; ''' 

request_5 = pd.io.sql.read_sql(query_5, con = engine)
request_5

avg_review
0   24.333333

Проверка средствами Python

In [ ]:
users = ratings.groupby('username')['rating_id'].count().reset_index()
users = users.query('rating_id > 50')
users

username  rating_id
57   jennifermiller         53
83       martinadam         56
104          paul88         56
110       richard89         55
123     sfitzgerald         55
153          xdavis         51

In [ ]:
users_lst = users["username"].tolist()
reviews.query('username in @users_lst').groupby('username')['review_id'].count().mean()


24.333333333333332

**Вывод:  среднее количество обзоров от пользователей, которые поставили больше 50 оценок равно  или 24.3(3)**

[*Назад к содержанию*](#Содержание)